optuna/trial/_trial.py

```diff
@@ -467,7 +466,7 @@ def objective(trial):
             )
             return
 
-        self.storage.set_trial_intermediate_value(self._trial_id, step, value)
+        self._cached_frozen_trial.intermediate_values[step] = value
```

Comment:
You must save the intermediate value to the storage and update the cache here.

```suggestion
        self.storage.set_trial_intermediate_value(self._trial_id, step, value)
        self._cached_frozen_trial.intermediate_values[step] = value
```

Reply:
Thank you very much for the suggestion.